sphinx/themes/basic/static/searchtools.js

```diff
@@ -514,8 +514,8 @@ const Search = {
     searchTerms.forEach((word) => {
       const files = [];
       const arr = [
-        { files: terms[word], score: Scorer.term },
-        { files: titleTerms[word], score: Scorer.title },
+        { files: terms.hasOwnProperty(word) ? terms[word] : [], score: Scorer.term },
```

Comment:
I'm actually a little confused by how this might work, as it doesn't appear that JavaScript supports setting the `__proto__` property in an object. Am I missing something? Should we be using a map to store these terms instead?

Reply:
When parsing JSON, an object key named `"__proto__"` is accepted, and converted to the JavaScript object key `["__proto__"]` -- distinct from the `__proto__` property (a reference to a class prototype).

During read-access, introducing `hasOwnProperty` allows us to filter out the latter, so that we don't mistakenly retrieve the prototype class reference from `terms`.

I'm not certain whether the below example will help to clarify, but I find experimentation helpful to understand some of the behaviours:

```javascript
> x = JSON.parse('{}')
{}
> y = JSON.parse('{"__proto__": "testing"}')
{ ['__proto__']: 'testing' }
> '__proto__' in x
true
> x.hasOwnProperty('__proto__')
false
> '__proto__' in y
true
> y.hasOwnProperty('__proto__')
true
```

Reply:
I forgot to respond to your `Map` question: that is certainly possible, although I feel that doing so might add some runtime overhead (since we'd need to load the native JS object contents from the index into respective `Map` objects).

As I understand it, `Map` objects are also more difficult to make immutable in JavaScript than `Object` and `Array` instances.  Freezing a `Map` object doesn't prevent its private contents from being updated:

```javascript
> x = {'foo': 'bar'}
{ foo: 'bar' }
> Object.freeze(x)
{ foo: 'bar' }
> x['foo'] = 'baz'
'baz'
> x['foo']
'bar'
> y = new Map([['foo', 'bar']])
Map(1) { 'foo' => 'bar' }
> Object.freeze(y)
Map(1) { 'foo' => 'bar' }
> y.set('foo', 'baz')
Map(1) { 'foo' => 'baz' }
> y.get('foo')
'baz'
```

So another factor is that, to the extent that I'd like to see the search index contents become immutable at load-time in future (#13098), I'd prefer not to migrate to the `Map` type unless there is a compelling reason to (maybe avoiding prototype pollution is that reason -- but I don't think we're exposed to that currently).

Reply:
Ok got it, this makes sense to me re: the property. TIL!

I'm a little bit less sure about whether the index should be immutable, but happy to defer that discussion to another day.

Reply:
Yep, OK - I'll try to add some phrasing to explain this succinctly.  Thanks for the code review!